core/vm/errors.go

```diff
@@ -40,8 +40,42 @@ var (
 	ErrGasUintOverflow          = errors.New("gas uint64 overflow")
 	ErrInvalidRetsub            = errors.New("invalid retsub")
 	ErrReturnStackExceeded      = errors.New("return stack limit reached")
-	ErrInvalidCode              = errors.New("invalid code")
 	ErrNonceUintOverflow        = errors.New("nonce uint64 overflow")
+	ErrInvalidCode              = errors.New("invalid code: must not begin with 0xef")
+	ErrInvalidEOFCode           = errors.New("invalid code: EOF validation failed")
+	ErrInvalidInterpreter       = errors.New("invalid interpreter")
```

Comment:
This new error has been added by me

Reply:
The need for this arises because erigon is still using the `Interpreter` interface, ostensibly to support external VMs. This was removed at ethereum/go-ethereum, which, with dropping configurable-EVM support, dropped the interface entirely. 

However, I'm not sure erigon's current approach is going to work with external VMs. 
Although I'm not sure, either, that they're actually intent on keeping these configuration options, since it appears EVMC support has been dropped too. 

So we can either
- aim to drop `type Interpreter interface` and have the `evm` only handle its native self, avoiding any type assertion here
- or adapt, somehow, the implementation so that it won't fail `ErrInvalidInterpreter` when the interpreter is not of the native type `*EVMInterpreter`.

Reply:
Nice write up. I reached the same outcome, though with the EOF PR, I was thinking that we want to deal with only the EOF surface without touching other stuff.
I suggest doing separate PRs for other changes.
What do you think?

Reply:
I think we should leave it as-is too, just making a review note of it when we do the PR.